### DISCUSSION TITLE
Weird behavior when building sys crate with multiple make jobs in docker container

### DISCUSSION THREAD

**Discussion by Christian-Prather** [corrosion-rs/corrosion#573] [Q&A] [answered]
Hello, 
I am running into a weird problem that I believe will go beyond just Corrosion however as I am only seeing it when using Corrosion I figured I would start here and see if anyone had any advice. Thank you in advance!

# Problem
The problem occurs when using corrosion to build a rust bin that relies on the `libffi-sys` crate https://crates.io/crates/libffi-sys/2.0.0. This external `libffi-sys` crate internally builds the `libffi` C library https://github.com/libffi/libffi using the cc crate https://crates.io/crates/cc as a dependency. The weirdness comes in the fact that libffi-sys fails to build under specific circumstances however this particular setup is critical for me to be able to use Corrosion. 

# Setup 
I am using the following build setup:
**Docker image based on Ubuntu 18.04**
**CMake version:** 3.30.5
**rustc version:** rustc 1.78.0 (9b00956e5 2024-04-29)

My test application can be found here but is essentially hello world bin with a basic cmake.
https://github.com/Christian-Prather/test-corrosion

Here are my commands I use to build the bin. 
```
cd test-corrosion
mkdir build && cd build
cmake -DCMAKE_BUILD_TYPE=Debug ..
make -j4
```
This will fail with the following log. 
[multiple-2.txt](https://github.com/user-attachments/files/17560353/multiple-2.txt)

However this is where it gets weird, if I run the same command without `-j4` so just `make` or `make -j1` it succeeds and produces and executable binary? I also am able to build with `make -j4` just fine if I am NOT in a docker container. 

Here is the log for `make -j1` inside the docker image which succeeds:
[single-2.txt](https://github.com/user-attachments/files/17560345/single-2.txt)

I know that since this only appears to occur in Docker that the problem likely does not fall entirely on corrosion however I was hoping anyone might have some thoughts on this. Unfortunately running in Docker for our builds is a hard requirement so even though it works without that's not an option. 

I am wondering if this has something to do with the make jobserver not being handled correctly in the Docker image however if that was true I would have assumed that the basic C++ program would of also failed with `make -jn; n > 1`? 

Any help at all would be really really appreciated as this seems to be the last hangup in porting a lot of CMake magic over to corrosion. Thank you,

# Additional Info:
Below is a table of test builds and their results
| Scenario    | Success / Fail build |
| -------- | ------- |
| Regular cargo build  | Works in and out of Docker image  |
| make -j8 | Works out of Docker image fails inside |
|make -j1  | Works in and out of Docker image  |
| Regular c++ project running make -j8 (parallel in my repo I posted) |  Works in and out of Docker image     |
| Building libffi C library directly |  Works in and out of Docker image      |
| Corrosion make -j8 without libffi-* crates |  Works in and out of Docker image     |

I have also tried setting cargo build flag if `-j to 1` and `codegen-units` to 1.

**Comment by jschwe** [comment, marked as answer]
>   make[5]: *** read jobs pipe: Resource temporarily unavailable.  Stop.

You use a top-level make (generated from cmake), which spawns a make job-server. 
Which make version do you have inside and outside of docker? I know something changed in make 4.4 with regards to the job server.
I believe the child make process from lib-ffi-sys will try to respect the make jobserver from the parent make, but it seems there is a problem there. This can be due to buggy make code, but I'm not a make expert, so you'll have to due some digging yourself to find out whats going wrong with the jobserver.

**Reply by Christian-Prather** [reply]
(Sorry was out of town) Ugh that was it, was on a version of make < 4.4 that worked a treat thank you once again!
